nixos/modules/services/misc/memos.nix

```
@@ -0,0 +1,176 @@
+{
+  config,
+  pkgs,
+  lib,
+  ...
+}:
+let
+  cfg = config.services.memos;
+in
+{
+  options.services.memos = with lib; {
+    enable = mkEnableOption "Memos note-taking";
+    package = mkPackageOption pkgs "memos" { };
+
+    openFirewall = mkEnableOption "opening the ports in the firewall";
+
+    user = mkOption {
+      type = types.str;
+      description = "The user to run memos as";
+      default = "memos";
+    };
+
+    group = mkOption {
+      type = types.str;
+      description = "The group to run memos as";
+      default = "memos";
+    };
+
+    mode = mkOption {
+      default = "prod";
+      type = types.enum [
+        "prod"
+        "dev"
+        "demo"
+      ];
+      description = ''
+        Sets the mode of the server, influencing its runtime behavior and the database used
+      '';
+    };
+    address = mkOption {
+      default = "0.0.0.0";
+      type = types.str;
+      description = ''
+        Specifies the address on which the server will listen for incoming connections
+      '';
+    };
+    port = mkOption {
+      default = 5230;
+      type = types.port;
+      description = ''
+        Sets the port on which the server will be accessible
+      '';
+    };
+    dataDir = mkOption {
+      default = "/var/lib/memos/";
+      type = types.path;
+      description = ''
+        Specifies the directory where Memos will store its data
+      '';
+    };
+    driver = mkOption {
+      default = "sqlite";
+      type = types.enum [
+        "sqlite"
+        "postgres"
+        "mysql"
+      ];
+      description = ''
+        Sets the database driver to be used by Memos
+      '';
+    };
+    dsn = mkOption {
+      default = null;
+      type = types.nullOr types.str;
+      description = ''
+        Specifies the database source name (DSN) for connecting to the database.
+        Note that each database driver has its own format for specifying the DSN.
+        See https://www.usememos.com/docs/install/database for more information.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    users.users.${cfg.user} = {
+      name = cfg.user;
+      description = lib.mkDefault "Memos service user";
+      isSystemUser = true;
+      group = cfg.group;
+    };
+
+    users.groups.${cfg.group} = lib.mapAttrs (name: lib.mkDefault) { };
```

```suggestion
    users.groups.${cfg.group} = { };
```
this is usually enough